### PR TITLE
Bundle Analysis: associate past assets to current parsed bundle

### PR DESCRIPTION
**Pull request by JerrySentry**
Asset name changes from one bundler build to another across commits, therefore we don't know if the current asset is a new one or a continuation of the previous one. This PR adds a function to offer a heuristic to associate the current asset to an asset from the previous commit.

The rules are that if the hashed asset name exists in the previous bundle report then this is considered the same asset. Similarly if all the modules of the asset are the same as any in the previous bundle report it is also considered the same asset. We track assets through a generated UUID, when 2 assets are considered associated then they will have the same UUID.

This mechanism is important for providing analytics asset size trends throughout the course of its existence. This component will be implemented in the coming iterations. 

### Legal Boilerplate

Look, I get it. The entity doing business as "Sentry" was incorporated in the State of Delaware in 2015 as Functional Software, Inc. In 2022 this entity acquired Codecov and as result Sentry is going to need some rights from me in order to utilize my contributions in this PR. So here's the deal: I retain all rights, title and interest in and to my contributions, and by keeping this boilerplate intact I confirm that Sentry can use, modify, copy, and redistribute my contributions, under Sentry's choice of terms.

### CODE REVIEW

File: shared/bundle_analysis/report.py [hunk 1]
```diff
@@ -170,6 +171,73 @@ def ingest(self, path: str) -> int:
...
+                            if curr_modules in prev_module_asset_mapping:
+                                associated_assets_found.append(
+                                    [
+                                        prev_module_asset_mapping[curr_modules],
+                                        curr_asset.uuid,
+                                    ]
+                                )
+
+                    # Update the Assets table for the bundle
+                    # TODO: Use SQLalchemy ORM to update instead of raw SQL
```
Comment (giovanni-guidini) [inline shared/bundle_analysis/report.py:234]: Is there a ticket for this TODO already? Big chances of it being forgotten

File: tests/unit/bundle_analysis/test_asset_association.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+def _get_asset_mapping(
+    bundle_analysis_report: BundleAnalysisReport, bundle_name: Asset
+) -> Dict[str, str]:
+    bundle_report = bundle_analysis_report.bundle_report(bundle_name)
+    asset_report = list(bundle_report.asset_reports())
+    return {asset.hashed_name: asset for asset in asset_report}
+
+
+def test_asset_association():
+    try:
```
Comment (giovanni-guidini) [inline tests/unit/bundle_analysis/test_asset_association.py:38]: Why have a `try:` without an `except`?

"to have the `finally;` block" 
but what's the difference of having it at the end of the try block?

**Review by giovanni-guidini — APPROVED** [review]
I personally think that `associate_previous_assets` is quite complex as it is. I'd encourage you to refactor it by breaking up parts of the big loop into helper functions.
Easier to read and to test. But the comments and docstrings are quite helpful, thanks for that.

Otherwise LGTM... left some other comments.

File: shared/bundle_analysis/report.py (hunk 1 above)
Comment (JerrySentry) [reply shared/bundle_analysis/report.py:234]: Added link to it.

File: shared/bundle_analysis/report.py [hunk 3]
```diff
@@ -170,6 +171,73 @@ def ingest(self, path: str) -> int:
         self.db_session.commit()
         return session_id
 
+    def associate_previous_assets(self, prev_bundle_analysis_report: Any) -> None:
+        """
+        Note: prev_bundle_analysis_report is of type BundleAnalysisReport,
+              typing.Self is not available in 3.10
```
[the remark this answers is not in the document]
Comment (JerrySentry) [reply shared/bundle_analysis/report.py]: Oh interesting I didn't know about the double quote trick!

**Review by giovanni-guidini — APPROVED** [review]

**Comment by JerrySentry** [comment]
> I personally think that `associate_previous_assets` is quite complex as it is. 

Separated out the 2 rules as 2 separate helper functions

File: shared/bundle_analysis/report.py [hunk 4]
```diff
@@ -170,6 +171,104 @@ def ingest(self, path: str) -> int:
         self.db_session.commit()
         return session_id
 
+    def _associate_bundle_report_assets_by_name(
+        self, curr_bundle_report, prev_bundle_report
+    ) -> Set:
```
Comment (giovanni-guidini) [inline shared/bundle_analysis/report.py:176]: [very nit] You can be more explicit by saying `Set[Tuple]` or even more `Set[Tuple[str, str]]`. Starts to get a bit wild fast but possible. I personally think `Set[Tuple]` is an improvement over `Set`.

File: shared/bundle_analysis/report.py [hunk 5]
```diff
@@ -170,6 +171,104 @@ def ingest(self, path: str) -> int:
         self.db_session.commit()
         return session_id
 
+    def _associate_bundle_report_assets_by_name(
+        self, curr_bundle_report, prev_bundle_report
```
Comment (giovanni-guidini) [inline shared/bundle_analysis/report.py:175]: The args don't have a known type?

File: shared/bundle_analysis/report.py [hunk 6]
```diff
@@ -170,6 +171,104 @@ def ingest(self, path: str) -> int:
...
+        Rule 1. Previous and current asset have the same hashed name
+        Rule 2. Previous and current asset shared the same set of module names
+        """
+        for curr_bundle_report in self.bundle_reports():
+            for prev_bundle_report in prev_bundle_analysis_report.bundle_reports():
+                if curr_bundle_report.name == prev_bundle_report.name:
+                    associated_assets_found = set()
+
+                    # Rule 1 check
+                    associated_assets_found |= (
```
Comment (giovanni-guidini) [inline shared/bundle_analysis/report.py:250]: Even the code is going "|=" (i.e. 😐)
(Jokes aside I do think this is an improvement, thanks)

**Review by giovanni-guidini — APPROVED** [review]

**Review by giovanni-guidini — APPROVED** [review]

**Pull request merged**